### PR TITLE
Command line interface

### PR DESCRIPTION
**Pull request by kaklakariada**
Create a command line interface providing commands `trace` and `convert`.

### CODE REVIEW

File: src/main/java/openfasttrack/cli/CliArguments.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public void setInputDir(final String inputDir)
+    {
+        this.inputDir = inputDir;
+    }
+
+    public Path getOutputFile()
+    {
+        if (this.outputFile == null)
+        {
+            throw new MissingArgumentException("outputFile");
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/CliArguments.java]: Make output file optional. Output to STDOUT per default.

File: src/main/java/openfasttrack/cli/CliArguments.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public void setUnnamedValues(final List<String> unnamedValues)
+    {
+        this.unnamedValues = unnamedValues;
+    }
+
+    public String getOutputFormat()
+    {
+        if (this.outputFormat == null)
+        {
+            throw new MissingArgumentException("outputFormat");
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/CliArguments.java]: Make argument optional. Use plain text per default.

File: src/main/java/openfasttrack/cli/CliArguments.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    private List<String> unnamedValues;
+    private String inputDir;
+    private String outputFile;
+    private String outputFormat;
+
+    public Path getInputDir()
+    {
+        if (this.inputDir == null)
+        {
+            throw new MissingArgumentException("inputDir");
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/CliArguments.java]: Make input file optional. Scan current working dir by default.

File: src/main/java/openfasttrack/cli/CommandLineInterpreter.java [hunk 4]
```diff
@@ -73,8 +76,11 @@ public CommandLineInterpreter(final String[] arguments, final Object argumentsRe
     {
         final Class<?> receiverClass = argumentsReceiver.getClass();
         final Stream<Method> methods = Arrays.stream(receiverClass.getMethods());
-        return methods.filter(method -> method.getName().startsWith(SETTER_PREFIX)) //
+        final Map<String, Method> setterMethods = methods
+                .filter(method -> method.getName().startsWith(SETTER_PREFIX)) //
                 .collect(toMap(CommandLineInterpreter::getSetterName, Function.identity()));
+        LOG.finest(() -> "Found arguments " + setterMethods.keySet());
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/CommandLineInterpreter.java]: Please rephrase. Log message sounds like parameter in the actual command entered by the user was found.

File: src/main/java/openfasttrack/cli/CommandLineInterpreter.java [hunk 5]
```diff
@@ -111,7 +117,8 @@ public void parse()
 
     private void handleNamedArgument(final ListIterator<String> iterator, final String argument)
     {
-        final String argumentName = argument.substring(SINGLE_CHAR_ARG_PREFIX.length());
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/CommandLineInterpreter.java]: I just realized that we don't have code to distinguish between single and multiple character parameters.

As per GNU convention the following is equivalent:

```
myprogram --verbose --recursive
myprogram -v -r
myprogram -vr
```

Wrote ticket #25.

File: src/main/java/openfasttrack/cli/MissingArgumentException.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+public class MissingArgumentException extends CliException
```
Comment (redcatbear) [inline src/main/java/openfasttrack/cli/MissingArgumentException.java]: Using exceptions to control the program flow here is a border case. Wrong user input is not really an exception but a regular situation. Nevertheless here it makes handling the code actually cleaner - so: approved.

File: src/main/java/openfasttrack/exporter/ExporterFactoryLoader.java [hunk 7]
```diff
@@ -29,11 +29,9 @@
 import java.util.stream.StreamSupport;
 
 import openfasttrack.core.ServiceLoaderWrapper;
-import openfasttrack.importer.ImporterException;
-import openfasttrack.importer.ImporterFactory;
 
 /**
- * This class is responsible for finding the matching {@link ImporterFactory}
+ * This class is responsible for finding the matching {@link ExporterFactory}
```
Comment (redcatbear) [inline src/main/java/openfasttrack/exporter/ExporterFactoryLoader.java]: Hah. Caught you! C+P error. ;-)

File: src/main/java/openfasttrack/importer/MultiFileImporter.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        final int itemCountBefore = this.specItemBuilder.getItemCount();
+        createImporter(file, DEFAULT_CHARSET, this.specItemBuilder).runImport();
+        final int itemCountImported = this.specItemBuilder.getItemCount() - itemCountBefore;
+        LOG.info(() -> "Imported " + itemCountImported + " items from '" + file + "'.");
+        return this;
+    }
+
+    /**
+     * Import all files from the given directory that match the given
+     * {@link FileSystem#getPathMatcher(String) glob expression} using character
```
Comment (redcatbear) [inline src/main/java/openfasttrack/importer/MultiFileImporter.java]: second "using" -> "and"

File: src/main/java/openfasttrack/importer/MultiFileImporter.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+     *            the dir to search for files to import.
+     * @param glob
+     *            the {@link FileSystem#getPathMatcher(String) glob expression}
+     * @return <code>this</code> for fluent programming style.
+     */
+    public MultiFileImporter importRecursiveDir(final Path dir, final String glob)
+    {
+        LOG.info(() -> "Importing files from '" + dir + "' matching glob '" + glob + "'...");
+        final PathMatcher matcher = dir.getFileSystem().getPathMatcher("glob:" + glob);
+        final AtomicInteger fileCount = new AtomicInteger(0);
```
Comment (redcatbear) [inline src/main/java/openfasttrack/importer/MultiFileImporter.java]: Introducing an AtomicInteger just for the same of producing a  log entry seems a little bit overdone to me. ;-)

File: src/main/java/openfasttrack/importer/specobject/ImportHelper.java [hunk 10]
```diff
@@ -74,12 +75,16 @@ private void foundStartElement(final StartElement element) throws XMLStreamExcep
...
+            {
+                throw new ImporterException("No specobject default doctype found");
+            }
             new SingleSpecobjectImportHelper(this.xmlEventReader, this.listener,
                     this.defaultDoctype).runImport();
             break;
 
         default:
-            LOG.warning(() -> "Found unknown start element " + element.getName());
+            LOG.warning(() -> "Found unknown xml element '" + element.getName() + "'");
```
Comment (redcatbear) [inline src/main/java/openfasttrack/importer/specobject/ImportHelper.java]: "XML"

File: src/main/java/openfasttrack/report/ReportService.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import openfasttrack.core.Trace;
+
+public class ReportService
```
Comment (redcatbear) [inline src/main/java/openfasttrack/report/ReportService.java]: Out of curiosity: Why did you create a service here?
My first instinct would have been to use an abstract factory to create a Reportable and then run it.

File: src/test/java/openfasttrack/cli/TestCliStarter.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        expectException(asList("convert"), MissingArgumentException.class,
+                "Argument 'inputDir' is missing");
+    }
+
+    @Test
+    public void testConvertUnknownExporter()
+    {
+        final Path outputFile = this.tempFolder.getRoot().toPath().resolve("output.xml");
+        expectException(
+                asList("convert", "-inputDir", this.docDir.toString(), "-outputFormat", "illegal",
```
Comment (redcatbear) [inline src/test/java/openfasttrack/cli/TestCliStarter.java]: See #25

File: src/main/java/openfasttrack/cli/CliArguments.java (hunk 2 above)
Comment (kaklakariada) [inline src/main/java/openfasttrack/cli/CliArguments.java]: Plain text is not yet supported for export, using `specobject` as default.

File: src/main/java/openfasttrack/importer/MultiFileImporter.java (hunk 9 above)
Comment (kaklakariada) [inline src/main/java/openfasttrack/importer/MultiFileImporter.java]: I know. But this is the easiest way to get the item count.

File: src/main/java/openfasttrack/report/ReportService.java (hunk 11 above)
Comment (kaklakariada) [inline src/main/java/openfasttrack/report/ReportService.java]: Because I was lazy. We should do this as soon as we have a second reporter format.

**Pull request merged**